### ISSUE TITLE
Abstract Layer Design

### ISSUE THREAD

**Issue by boolafish** [omisego/research#60]
- A subtask of #49 
- Goal: 
  - Able to support ODEX tx
  - Able to upgrade our ODEX in the future
  - Try to be flexible for adding new features on top of plasma

**Comment by boolafish** [comment]
Started from the discussion in #59, trying to support the need there, I changed my initial design and try to break the "predicate" of whole app into "rules" that can be combined to provide all logic needed for ODEX.

Below is my take-1 attempt.
https://omisego.atlassian.net/wiki/spaces/RES/pages/32014470/UTXO-Based+Rule+Design

TL;DR: introduce “rule” that binds with data/state. And all interactive game start with disclosing data assuming its correctness. If invalid, would be challenged with a certain rule.

**Comment by boolafish** [comment]
Abstraction layer design PR:

TL;DR Bind tx type to exit game, and define how to exit there. So we can add tx type for new features on top of plasma.

https://github.com/omisego/plasma-contracts/pull/93

**Comment by boolafish** [comment]
[Notes] Writing down some notes from the calls with @paulperegud 

- It might be possible we only need to check two things: 1. sum of inputs >= sum of outputs 2. output is used (correctly) or not
- Aside from the tx output owner, other people might not really care whether the tx is totally correct or not as long as the previous two things hold. 

What this implies? Probably we can have predicate that checks only the above two statement. Watcher can use that predicate for output to check things they need. Exchange can have customized watcher that can check trade settlement tx correctness.

**Comment by boolafish** [comment]
Did a review on the design: [doc here](https://omisego.atlassian.net/wiki/spaces/RES/pages/35520517/UTXO+Based+Plasma+Abstract+Layer+Design)

And the meeting note: [here](https://omisego.atlassian.net/wiki/spaces/RES/pages/36962323/2019-04-26+Abstract+Layer+Design+Review+Meeting+notes)

Have follow up action items and create the POC implementation stories. Should follow those stories then.
- POC implementation story: https://github.com/omisego/research/issues/78 
- Watcher Predicate minimum responsibility: https://github.com/omisego/research/issues/79
